d2/src/test/java/com/linkedin/d2/balancer/strategies/relative/StateUpdaterTest.java

```diff
@@ -0,0 +1,359 @@
+  public void testHealthScoreRecover(double currentHealthScore, double slowStartThreshold)
+  {
+    List<TrackerClient> trackerClients = TrackerClientMockHelper.mockTrackerClients(3,
+        Arrays.asList(20, 20, 20), Arrays.asList(0, 0, 0), Arrays.asList(300L, 300L, 300L),
+        Arrays.asList(200L, 200L, 200L), Arrays.asList(0, 0, 0));
+
+    PartitionState state = new PartitionStateDataBuilder(new DistributionNonDiscreteRingFactory<>())
+        .setClusterGenerationId(DEFAULT_CLUSTER_GENERATION_ID)
+        .setTrackerClientStateMap(trackerClients, Arrays.asList(currentHealthScore, 1.0, 1.0),
+            Arrays.asList(TrackerClientState.HealthState.UNHEALTHY, TrackerClientState.HealthState.HEALTHY, TrackerClientState.HealthState.HEALTHY),
+            Arrays.asList(20, 20, 20), RelativeLoadBalancerStrategyFactory.DEFAULT_INITIAL_HEALTH_SCORE,
+            RelativeLoadBalancerStrategyFactory.DEFAULT_MIN_CALL_COUNT)
+        .build();
+
+    ConcurrentMap<Integer, PartitionState> partitionLoadBalancerStateMap = new ConcurrentHashMap<>();
+    partitionLoadBalancerStateMap.put(DEFAULT_PARTITION_ID, state);
+    setup(new D2RelativeStrategyProperties().setSlowStartThreshold(slowStartThreshold),
+        partitionLoadBalancerStateMap);
+
+    _stateUpdater.updateState();
+    Map<URI, Integer> pointsMap = _stateUpdater.getPointsMap(DEFAULT_PARTITION_ID);
+
+    if (slowStartThreshold == RelativeLoadBalancerStrategyFactory.DEFAULT_SLOW_START_THRESHOLD)
+    {
+      assertEquals(pointsMap.get(trackerClients.get(0).getUri()).intValue(), 5);
+      assertEquals(pointsMap.get(trackerClients.get(1).getUri()).intValue(), HEALTHY_POINTS);
+      assertEquals(pointsMap.get(trackerClients.get(2).getUri()).intValue(), HEALTHY_POINTS);
+    } else if (currentHealthScore == 0.0)
+    {
+      assertEquals(pointsMap.get(trackerClients.get(0).getUri()).intValue(), 1);
+      assertEquals(pointsMap.get(trackerClients.get(1).getUri()).intValue(), HEALTHY_POINTS);
+      assertEquals(pointsMap.get(trackerClients.get(2).getUri()).intValue(), HEALTHY_POINTS);
+    } else if (currentHealthScore == 0.1)
+    {
+      assertEquals(pointsMap.get(trackerClients.get(0).getUri()).intValue(), 20);
+      assertEquals(pointsMap.get(trackerClients.get(1).getUri()).intValue(), HEALTHY_POINTS);
+      assertEquals(pointsMap.get(trackerClients.get(2).getUri()).intValue(), HEALTHY_POINTS);
+    } else
+    {
+      assertEquals(pointsMap.get(trackerClients.get(0).getUri()).intValue(), 30);
+      assertEquals(pointsMap.get(trackerClients.get(1).getUri()).intValue(), HEALTHY_POINTS);
+      assertEquals(pointsMap.get(trackerClients.get(2).getUri()).intValue(), HEALTHY_POINTS);
+    }
+  }
+
+  @DataProvider(name = "slowStartThreshold")
+  Object[][] getSlowStartThreshold()
+  {
+    return new Object[][]
+        {
+            {0.0, 0.0},
+            {0.0, 0.2},
+            {0.1, 0.2},
+            {0.25, 0.2}
+        };
+  }
+}
```

Comment:
add test that verifies the executor is scheduled to run updateState periodically.